File: go/internal/changefile/changed_packages_test.go
package changefile_test

import (
	"slices"
	"testing"

	"github.com/microsoft/beachball/internal/changefile"
	"github.com/microsoft/beachball/internal/monorepo"
	"github.com/microsoft/beachball/internal/options"
	"github.com/microsoft/beachball/internal/testutil"
	"github.com/microsoft/beachball/internal/types"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// get default options for this file (fetch disabled)
func getDefaultOptions() types.BeachballOptions {
	defaultOptions := types.DefaultOptions()
	defaultOptions.Branch = testutil.DefaultRemoteBranch
	defaultOptions.Fetch = false

	return defaultOptions
}

func getOptionsAndPackages(t *testing.T, repo *testutil.Repository, overrides *types.BeachballOptions, extraCli *types.CliOptions) (types.BeachballOptions, types.PackageInfos, types.ScopedPackages) {
	t.Helper()

	cli := types.CliOptions{}
	if extraCli != nil {
		cli = *extraCli
	}

	repoOpts := getDefaultOptions()
	if overrides != nil {
		repoOpts = *overrides
	}

	parsed := options.GetParsedOptionsForTest(repo.RootPath(), cli, repoOpts)
	packageInfos, err := monorepo.GetPackageInfos(&parsed.Options)
	require.NoError(t, err, "failed to get package infos")
	scopedPackages := monorepo.GetScopedPackages(&parsed.Options, packageInfos)
	return parsed.Options, packageInfos, scopedPackages
}

func checkOutTestBranch(repo *testutil.Repository, name string) {
	repo.Checkout("-b", name, testutil.DefaultBranch)
}

// ===== Basic tests (TS: getChangedPackages (basic)) =====

// TS: "returns empty list when no changes have been made"
func TestReturnsEmptyListWhenNoChanges(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()

	opts, infos, scoped := getOptionsAndPackages(t, repo, nil, nil)
	result, err := changefile.GetChangedPackages(&opts, infos, scoped)
	require.NoError(t, err)
	assert.Empty(t, result)
}

// TS: "returns package name when changes exist in a new branch"
func TestReturnsPackageNameWhenChangesInBranch(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()
	checkOutTestBranch(repo, "changes_in_branch")
	repo.CommitChange("packages/foo/myFilename")

	buf := testutil.CaptureLogging(t)
	overrides := getDefaultOptions()
	overrides.Verbose = true
	opts, infos, scoped := getOptionsAndPackages(t, repo, &overrides, nil)
	result, err := changefile.GetChangedPackages(&opts, infos, scoped)
	require.NoError(t, err)
	assert.Equal(t, []string{"foo"}, result)
	assert.Contains(t, buf.String(), "Checking for changes against")
}

// TS: "returns empty list when changes are CHANGELOG files"
func TestReturnsEmptyListForChangelogChanges(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()
	checkOutTestBranch(repo, "changelog_changes")
	repo.CommitChange("packages/foo/CHANGELOG.md")

	opts, infos, scoped := getOptionsAndPackages(t, repo, nil, nil)
	result, err := changefile.GetChangedPackages(&opts, infos, scoped)
	require.NoError(t, err)
	assert.Empty(t, result)
}

// TS: "returns the given package name(s) as-is"
func TestReturnsGivenPackageNamesAsIs(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()

	cli := types.CliOptions{Package: []string{"foo"}}
	opts, infos, scoped := getOptionsAndPackages(t, repo, nil, &cli)
	result, err := changefile.GetChangedPackages(&opts, infos, scoped)
	require.NoError(t, err)
	assert.Equal(t, []string{"foo"}, result)

	cli2 := types.CliOptions{Package: []string{"foo", "bar", "nope"}}
	opts2, infos2, scoped2 := getOptionsAndPackages(t, repo, nil, &cli2)
	result2, err := changefile.GetChangedPackages(&opts2, infos2, scoped2)
	require.NoError(t, err)
	assert.Equal(t, []string{"foo", "bar", "nope"}, result2)
}

// TS: "returns all packages with all: true"
func TestReturnsAllPackagesWithAllTrue(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()

	overrides := getDefaultOptions()
	overrides.All = true
	opts, infos, scoped := getOptionsAndPackages(t, repo, &overrides, nil)
	result, err := changefile.GetChangedPackages(&opts, infos, scoped)
	require.NoError(t, err)
	slices.Sort(result)
	assert.Equal(t, []string{"a", "b", "bar", "baz", "foo"}, result)
}

// ===== Single package tests (TS: getChangedPackages) =====

// TS: "detects changed files in single-package repo"
func TestDetectsChangedFilesInSinglePackageRepo(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "single")
	repo := factory.CloneRepository()

	opts, infos, scoped := getOptionsAndPackages(t, repo, nil, nil)
	result, err := changefile.GetChangedPackages(&opts, infos, scoped)
	require.NoError(t, err)
	assert.Empty(t, result)

	repo.StageChange("foo.js")
	result2, err := changefile.GetChangedPackages(&opts, infos, scoped)
	require.NoError(t, err)
	assert.Equal(t, []string{"foo"}, result2)
}

// TS: "respects ignorePatterns option"
func TestRespectsIgnorePatterns(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "single")
	repo := factory.CloneRepository()

	overrides := getDefaultOptions()
	overrides.IgnorePatterns = []string{"*.test.js", "tests/**", "yarn.lock"}
	overrides.Verbose = true

	opts, infos, scoped := getOptionsAndPackages(t, repo, &overrides, nil)

	repo.WriteFile("src/foo.test.js")
	repo.WriteFile("tests/stuff.js")
	repo.WriteFileContent("yarn.lock", "changed")
	repo.Git([]string{"add", "-A"})

	buf := testutil.CaptureVerboseLogging(t)
	result, err := changefile.GetChangedPackages(&opts, infos, scoped)
	require.NoError(t, err)
	assert.Empty(t, result)
	output := buf.String()
	assert.Contains(t, output, "ignored by pattern")
	assert.Contains(t, output, "All files were ignored")
}

// ===== Monorepo tests =====

// TS: "detects changed files in monorepo"
func TestDetectsChangedFilesInMonorepo(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()

	opts, infos, scoped := getOptionsAndPackages(t, repo, nil, nil)
	result, err := changefile.GetChangedPackages(&opts, infos, scoped)
	require.NoError(t, err)
	assert.Empty(t, result)

	repo.StageChange("packages/foo/test.js")
	result2, err := changefile.GetChangedPackages(&opts, infos, scoped)
	require.NoError(t, err)
	assert.Equal(t, []string{"foo"}, result2)
}

// TS: "excludes packages that already have change files"
func TestExcludesPackagesWithExistingChangeFiles(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()
	repo.Checkout("-b", "test")
	repo.CommitChange("packages/foo/test.js")

	overrides := getDefaultOptions()
	overrides.Verbose = true
	opts, infos, scoped := getOptionsAndPackages(t, repo, &overrides, nil)
	testutil.GenerateChangeFiles(t, []string{"foo"}, &opts, repo)

	buf := testutil.CaptureVerboseLogging(t)
	result, err := changefile.GetChangedPackages(&opts, infos, scoped)
	require.NoError(t, err)
	assert.Empty(t, result)
	output := buf.String()
	assert.Contains(t, output, "already has change files for these packages")
	assert.Contains(t, output, "Found 1 file in 1 package that should be published")

	// Change bar => bar is the only changed package returned
	repo.StageChange("packages/bar/test.js")
	result2, err := changefile.GetChangedPackages(&opts, infos, scoped)
	require.NoError(t, err)
	assert.Equal(t, []string{"bar"}, result2)
}

// TS: "ignores change files that exist in target remote branch"
func TestIgnoresChangeFilesThatExistInTargetRemoteBranch(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "single")
	repo := factory.CloneRepository()

	overrides := getDefaultOptions()
	overrides.Verbose = true
	opts, infos, scoped := getOptionsAndPackages(t, repo, &overrides, nil)

	// create and push a change file in master
	testutil.GenerateChangeFiles(t, []string{"foo"}, &opts, repo)
	repo.Push()

	// create a new branch and stage a new file
	repo.Checkout("-b", "test")
	repo.StageChange("test.js")

	buf := testutil.CaptureVerboseLogging(t)
	result, err := changefile.GetChangedPackages(&opts, infos, scoped)
	require.NoError(t, err)
	assert.Equal(t, []string{"foo"}, result)
	// The change file from master should not appear in the exclusion list
	assert.NotContains(t, buf.String(), "already has change files")
}

// TS: "ignores package changes as appropriate"
func TestIgnoresPackageChangesAsAppropriate(t *testing.T) {
	rootPkg := map[string]any{
		"name":       "test-monorepo",
		"version":    "1.0.0",
		"private":    true,
		"workspaces": []string{"packages/*"},
	}

	packages := map[string]map[string]any{
		"private-pkg": {"name": "private-pkg", "version": "1.0.0", "private": true},
		"no-publish": {
			"name": "no-publish", "version": "1.0.0",
			"beachball": map[string]any{"shouldPublish": false},
		},
		"out-of-scope": {"name": "out-of-scope", "version": "1.0.0"},
		"ignore-pkg":   {"name": "ignore-pkg", "version": "1.0.0"},
		"publish-me":   {"name": "publish-me", "version": "1.0.0"},
	}

	groups := map[string]map[string]map[string]any{
		"packages": packages,
	}

	factory := testutil.NewCustomRepositoryFactory(t, rootPkg, groups)
	repo := factory.CloneRepository()

	repo.StageChange("packages/private-pkg/test.js")
	repo.StageChange("packages/no-publish/test.js")
	repo.StageChange("packages/out-of-scope/test.js")
	repo.StageChange("packages/ignore-pkg/jest.config.js")
	repo.StageChange("packages/ignore-pkg/CHANGELOG.md")
	repo.StageChange("packages/publish-me/test.js")

	overrides := getDefaultOptions()
	overrides.Scope = []string{"!packages/out-of-scope"}
	overrides.IgnorePatterns = []string{"**/jest.config.js"}
	overrides.Verbose = true

	buf := testutil.CaptureVerboseLogging(t)
	opts, infos, scoped := getOptionsAndPackages(t, repo, &overrides, nil)
	result, err := changefile.GetChangedPackages(&opts, infos, scoped)
	require.NoError(t, err)
	assert.Equal(t, []string{"publish-me"}, result)
	output := buf.String()
	assert.Contains(t, output, "private-pkg is private")
	assert.Contains(t, output, "no-publish has beachball.shouldPublish=false")
	assert.Contains(t, output, "out-of-scope is out of scope")
	assert.Contains(t, output, "ignored by pattern")
}

// TS: "detects changed files in multi-root monorepo repo"
func TestDetectsChangedFilesInMultiRootMonorepo(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "multi-project")
	repo := factory.CloneRepository()

	repo.StageChange("project-a/packages/foo/test.js")

	// Test from project-a root
	pathA := repo.PathTo("project-a")
	optsA := getDefaultOptions()
	optsA.Path = pathA

	parsedA := options.GetParsedOptionsForTest(pathA, types.CliOptions{}, optsA)
	infosA, err := monorepo.GetPackageInfos(&parsedA.Options)
	require.NoError(t, err, "failed to get package infos")
	scopedA := monorepo.GetScopedPackages(&parsedA.Options, infosA)
	resultA, err := changefile.GetChangedPackages(&parsedA.Options, infosA, scopedA)
	require.NoError(t, err)
	assert.Equal(t, []string{"@project-a/foo"}, resultA)

	// Test from project-b root
	pathB := repo.PathTo("project-b")
	optsB := getDefaultOptions()
	optsB.Path = pathB

	parsedB := options.GetParsedOptionsForTest(pathB, types.CliOptions{}, optsB)
	infosB, err := monorepo.GetPackageInfos(&parsedB.Options)
	require.NoError(t, err, "failed to get package infos")
	scopedB := monorepo.GetScopedPackages(&parsedB.Options, infosB)
	resultB, err := changefile.GetChangedPackages(&parsedB.Options, infosB, scopedB)
	require.NoError(t, err)
	assert.Empty(t, resultB)
}

File: rust/tests/changed_packages_test.rs
mod common;

use beachball::changefile::changed_packages::get_changed_packages;
use beachball::monorepo::package_infos::get_package_infos;
use beachball::monorepo::scoped_packages::get_scoped_packages;
use beachball::options::get_options::get_parsed_options_for_test;
use beachball::types::options::{BeachballOptions, CliOptions};
use common::change_files::generate_change_files;
use common::repository::Repository;
use common::repository_factory::RepositoryFactory;
use std::collections::HashMap;

const DEFAULT_BRANCH: &str = "master";
const DEFAULT_REMOTE_BRANCH: &str = "origin/master";

fn get_options_and_packages(
    repo: &Repository,
    overrides: Option<BeachballOptions>,
    extra_cli: Option<CliOptions>,
) -> (BeachballOptions, beachball::types::package_info::PackageInfos, beachball::types::package_info::ScopedPackages) {
    let cli = extra_cli.unwrap_or_default();
    let mut repo_opts = overrides.unwrap_or_default();
    repo_opts.branch = DEFAULT_REMOTE_BRANCH.to_string();
    repo_opts.fetch = false;

    let parsed = get_parsed_options_for_test(repo.root_path(), cli, repo_opts);
    let package_infos = get_package_infos(&parsed.options).unwrap();
    let scoped_packages = get_scoped_packages(&parsed.options, &package_infos);
    (parsed.options, package_infos, scoped_packages)
}

fn check_out_test_branch(repo: &Repository, name: &str) {
    let branch_name = name.replace(|c: char| !c.is_alphanumeric(), "-");
    repo.checkout(&["-b", &branch_name, DEFAULT_BRANCH]);
}

// ===== Basic tests =====

#[test]
fn returns_empty_list_when_no_changes() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();

    let (options, infos, scoped) = get_options_and_packages(&repo, None, None);
    let result = get_changed_packages(&options, &infos, &scoped).unwrap();
    assert!(result.is_empty());
}

#[test]
fn returns_package_name_when_changes_in_branch() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();
    check_out_test_branch(&repo, "changes_in_branch");
    repo.commit_change("packages/foo/myFilename");

    let (options, infos, scoped) = get_options_and_packages(&repo, None, None);
    let result = get_changed_packages(&options, &infos, &scoped).unwrap();
    assert_eq!(result, vec!["foo"]);
}

#[test]
fn returns_empty_list_for_changelog_changes() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();
    check_out_test_branch(&repo, "changelog_changes");
    repo.commit_change("packages/foo/CHANGELOG.md");

    let (options, infos, scoped) = get_options_and_packages(&repo, None, None);
    let result = get_changed_packages(&options, &infos, &scoped).unwrap();
    assert!(result.is_empty());
}

#[test]
fn returns_given_package_names_as_is() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();

    let mut cli = CliOptions::default();
    cli.package = Some(vec!["foo".to_string()]);
    let (options, infos, scoped) = get_options_and_packages(&repo, None, Some(cli));
    let result = get_changed_packages(&options, &infos, &scoped).unwrap();
    assert_eq!(result, vec!["foo"]);

    let mut cli2 = CliOptions::default();
    cli2.package = Some(vec!["foo".to_string(), "bar".to_string(), "nope".to_string()]);
    let (options2, infos2, scoped2) = get_options_and_packages(&repo, None, Some(cli2));
    let result2 = get_changed_packages(&options2, &infos2, &scoped2).unwrap();
    assert_eq!(result2, vec!["foo", "bar", "nope"]);
}

#[test]
fn returns_all_packages_with_all_true() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();

    let mut opts = BeachballOptions::default();
    opts.all = true;
    let (options, infos, scoped) = get_options_and_packages(&repo, Some(opts), None);
    let mut result = get_changed_packages(&options, &infos, &scoped).unwrap();
    result.sort();
    assert_eq!(result, vec!["a", "b", "bar", "baz", "foo"]);
}

// ===== Single package tests =====

#[test]
fn detects_changed_files_in_single_package_repo() {
    let factory = RepositoryFactory::new("single");
    let repo = factory.clone_repository();

    let (options, infos, scoped) = get_options_and_packages(&repo, None, None);
    assert!(get_changed_packages(&options, &infos, &scoped).unwrap().is_empty());

    repo.stage_change("foo.js");
    let result = get_changed_packages(&options, &infos, &scoped).unwrap();
    assert_eq!(result, vec!["foo"]);
}

#[test]
fn respects_ignore_patterns() {
    let factory = RepositoryFactory::new("single");
    let repo = factory.clone_repository();

    let mut opts = BeachballOptions::default();
    opts.ignore_patterns = Some(vec![
        "*.test.js".to_string(),
        "tests/**".to_string(),
        "yarn.lock".to_string(),
    ]);
    opts.verbose = true;

    let (options, infos, scoped) = get_options_and_packages(&repo, Some(opts), None);

    repo.write_file("src/foo.test.js");
    repo.write_file("tests/stuff.js");
    repo.write_file_content("yarn.lock", "changed");
    repo.git(&["add", "-A"]);

    let result = get_changed_packages(&options, &infos, &scoped).unwrap();
    assert!(result.is_empty());
}

// ===== Monorepo tests =====

#[test]
fn detects_changed_files_in_monorepo() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();

    let (options, infos, scoped) = get_options_and_packages(&repo, None, None);
    assert!(get_changed_packages(&options, &infos, &scoped).unwrap().is_empty());

    repo.stage_change("packages/foo/test.js");
    let result = get_changed_packages(&options, &infos, &scoped).unwrap();
    assert_eq!(result, vec!["foo"]);
}

#[test]
fn excludes_packages_with_existing_change_files() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();
    repo.checkout(&["-b", "test"]);
    repo.commit_change("packages/foo/test.js");

    let mut opts = BeachballOptions::default();
    opts.verbose = true;
    let (options, infos, scoped) = get_options_and_packages(&repo, Some(opts), None);
    generate_change_files(&["foo"], &options, &repo);

    let result = get_changed_packages(&options, &infos, &scoped).unwrap();
    assert!(result.is_empty(), "Expected empty but got: {result:?}");

    // Change bar => bar is the only changed package returned
    repo.stage_change("packages/bar/test.js");
    let result2 = get_changed_packages(&options, &infos, &scoped).unwrap();
    assert_eq!(result2, vec!["bar"]);
}

#[test]
fn ignores_package_changes_as_appropriate() {
    use serde_json::json;

    let mut packages: HashMap<String, serde_json::Value> = HashMap::new();
    packages.insert("private-pkg".to_string(), json!({
        "name": "private-pkg", "version": "1.0.0", "private": true
    }));
    packages.insert("no-publish".to_string(), json!({
        "name": "no-publish", "version": "1.0.0",
        "beachball": { "shouldPublish": false }
    }));
    packages.insert("out-of-scope".to_string(), json!({
        "name": "out-of-scope", "version": "1.0.0"
    }));
    packages.insert("ignore-pkg".to_string(), json!({
        "name": "ignore-pkg", "version": "1.0.0"
    }));
    packages.insert("publish-me".to_string(), json!({
        "name": "publish-me", "version": "1.0.0"
    }));

    let root = json!({
        "name": "test-monorepo",
        "version": "1.0.0",
        "private": true,
        "workspaces": ["packages/*"]
    });

    let factory = RepositoryFactory::new_custom(root, vec![("packages".to_string(), packages)]);
    let repo = factory.clone_repository();

    repo.stage_change("packages/private-pkg/test.js");
    repo.stage_change("packages/no-publish/test.js");
    repo.stage_change("packages/out-of-scope/test.js");
    repo.stage_change("packages/ignore-pkg/jest.config.js");
    repo.stage_change("packages/ignore-pkg/CHANGELOG.md");
    repo.stage_change("packages/publish-me/test.js");

    let mut opts = BeachballOptions::default();
    opts.scope = Some(vec!["!packages/out-of-scope".to_string()]);
    opts.ignore_patterns = Some(vec!["**/jest.config.js".to_string()]);
    opts.verbose = true;

    let (options, infos, scoped) = get_options_and_packages(&repo, Some(opts), None);
    let result = get_changed_packages(&options, &infos, &scoped).unwrap();
    assert_eq!(result, vec!["publish-me"]);
}

#[test]
fn detects_changed_files_in_multi_root_monorepo() {
    let factory = RepositoryFactory::new("multi-project");
    let repo = factory.clone_repository();

    repo.stage_change("project-a/packages/foo/test.js");

    // Test from project-a root
    let path_a = repo.path_to(&["project-a"]).to_string_lossy().to_string();
    let mut opts_a = BeachballOptions::default();
    opts_a.path = path_a.clone();
    opts_a.branch = DEFAULT_REMOTE_BRANCH.to_string();
    opts_a.fetch = false;

    let parsed_a = get_parsed_options_for_test(
        &path_a,
        CliOptions::default(),
        opts_a,
    );
    let infos_a = get_package_infos(&parsed_a.options).unwrap();
    let scoped_a = get_scoped_packages(&parsed_a.options, &infos_a);
    let result_a = get_changed_packages(&parsed_a.options, &infos_a, &scoped_a).unwrap();
    assert_eq!(result_a, vec!["@project-a/foo"]);

    // Test from project-b root
    let path_b = repo.path_to(&["project-b"]).to_string_lossy().to_string();
    let mut opts_b = BeachballOptions::default();
    opts_b.path = path_b.clone();
    opts_b.branch = DEFAULT_REMOTE_BRANCH.to_string();
    opts_b.fetch = false;

    let parsed_b = get_parsed_options_for_test(
        &path_b,
        CliOptions::default(),
        opts_b,
    );
    let infos_b = get_package_infos(&parsed_b.options).unwrap();
    let scoped_b = get_scoped_packages(&parsed_b.options, &infos_b);
    let result_b = get_changed_packages(&parsed_b.options, &infos_b, &scoped_b).unwrap();
    assert!(result_b.is_empty());
}
